pinot-spi/src/main/java/org/apache/pinot/spi/data/function/evaluators/DefaultTimeSpecEvaluator.java

```diff
@@ -0,0 +1,72 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.data.function.evaluators;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import java.util.List;
+import org.apache.pinot.spi.data.TimeFieldSpec;
+import org.apache.pinot.spi.data.TimeGranularitySpec;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.utils.TimeConverter;
+import org.apache.pinot.spi.utils.TimeUtils;
+
+
+/**
+ * The {@code DefaultTimeSpecEvaluator} class will convert the time value based on the {@link TimeFieldSpec}.
+ */
+public class DefaultTimeSpecEvaluator implements ExpressionEvaluator {
+  private String _incomingTimeColumn;
+  private String _outgoingTimeColumn;
+  private TimeConverter _incomingTimeConverter;
+  private TimeConverter _outgoingTimeConverter;
+  private boolean _isValidated = false;
+
+  public DefaultTimeSpecEvaluator(TimeGranularitySpec incomingGranularitySpec,
+      TimeGranularitySpec outgoingGranularitySpec) {
+    Preconditions.checkState(!incomingGranularitySpec.equals(outgoingGranularitySpec));
+    _incomingTimeColumn = incomingGranularitySpec.getName();
+    _outgoingTimeColumn = outgoingGranularitySpec.getName();
+    Preconditions.checkState(!_incomingTimeColumn.equals(_outgoingTimeColumn));
+    _incomingTimeConverter = new TimeConverter(incomingGranularitySpec);
+    _outgoingTimeConverter = new TimeConverter(outgoingGranularitySpec);
+  }
+
+  @Override
+  public List<String> getArguments() {
+    return Lists.newArrayList(_incomingTimeColumn, _outgoingTimeColumn);
```

Comment:
Argument should only contain `_incomingTimeColumn`?

Reply:
No. If it contains only incoming, we will never read the outgoing in the record extractor. The ExpressionTransformer will not know about outgoing time value. This is necessary to handle the case where outgoing is already computed by user.

Reply:
I don't follow. Arguments should only contain the input columns. For time conversion, the input should always be the incoming time column, it should never be the outgoing time column.
The fields passed to the record extractor should be the outgoing column name, and the arguments for the evaluator should be the incoming column name right?

Reply:
Based on offline discussion, added support for case where transformed destination field already exists in the data.
DefaultTimeSpecEvaluator looks right now i.e. only incoming in getArguments.